### ISSUE TITLE
Going off hook is often not happening on USR 5637

### ISSUE THREAD

**Issue by whitedavidp** [thess/callattendant#11]
I finally received my new (used) USR 5637 and have it hooked up. I have already flashed the most recent firmware and it appears to support voice activity.

And, indeed, things do seem to work for the most part. However, many times when a call is screened (I cannot yet see a pattern to it), the log shows "Going off hook..." but the phone keeps ringing and ringing - at least until I hang up or the call goes to my local telecom's voice mail system. And then I start seeing what I expect to see in the log. Stuff like:

Jan 16 14:26:26 callattendant[3548]: RING LED blinking: 10 times
Jan 16 14:26:32 callattendant[3548]: >> Playing greeting...
Jan 16 14:26:36 callattendant[3548]: >> Starting voice mail...
Jan 16 14:26:36 callattendant[3548]: MSG LED blinking
Jan 16 14:26:36 callattendant[3548]: > Waiting for key-press...
Jan 16 14:26:45 callattendant[3548]: Timeout limit exceeded: 8
Jan 16 14:26:45 callattendant[3548]: Exiting wait_for_keypress(8): Timeout - wait time limit reached.
Jan 16 14:26:45 callattendant[3548]: MSG LED turned off
Jan 16 14:26:45 callattendant[3548]: MSG count: 0
Jan 16 14:26:45 callattendant[3548]: > Going on hook...
Jan 16 14:26:45 callattendant[3548]: Waiting for next call...

Sometimes this actually works. I can hear a greeting and even record a voice message and play it back via the web UI. But I would say that the majority of the time, I encounter this mis-behavior.

Here are the screening variables from my config file:

SCREENED_ACTIONS = ("answer", "greeting", "voice_mail")
SCREENED_GREETING_FILE = "general_greeting.wav"
SCREENED_RINGS_BEFORE_ANSWER = 0

Any suggestions?

**Comment by thess** [comment]
I'm not sure what you expect to have happen here. The log output supplied looks normal to me for 'screening' action. One problem I can see here is related to the messed up configuration of WAV files in the app. The correct WAV file for screening is "screener_greeting.wav" which you don't have at the moment. This was my fault for not checking-in my WAV files that match the 'screening' behavior as currently intended. Sorry for that.

The addition of the correct WAV files and changes to installation and configuration should be in the repo tomorrow. A new release shortly thereafter if you want a Python 'wheel'.

**Comment by whitedavidp** [comment]
Thanks, as always, for your time and help!

I agree that everything "looks" right in the log I attached. I will have to try to do better, if I can today. But what is happening is that the log shows "going OFF HOOK", but that does not seem to happen with the modem even though it says so. The phone continues to ring and ring - which should not happen if the modem has indeed gone off-hook, right?

No worries about the .wav file. I really am planning to use one that says "Your call is being screened. If you are a human, press 1 to be temporarily unblocked, hang up, and immediately call back" before transferring all screened calls to voice_mail where I will look for the key press and (hopefully - but still not quite sure how) this will result in the activation (on a timeout of maybe 2 minutes) of the Permit Next Call mechanism.

It seems to me that the existing Permit Next Call mechanism does NOT have a timeout. It simply sets the semaphore and leaves it there for however long it takes for another call to come in. I am not sure that makes great sense in any use case I can think of. So I think it would be great if there was a configurable timeout for Permit Next Call in minutes and that it always be applied. Of course, if set to zero, I guess this could also be ignored for those who don't want it. I am not sure I can do any of this on my own given my limited python skills.

Thanks

**Comment by thess** [comment]
I'd like to see more log output - especially the output from the modem in response to commands sent. If you set `DEBUG = True` in your `app.cfg` settings, all modem exchanges will be logged. Also, please don't truncate the output you are attaching to a problem report. I want to see what output was logged before the start of what you submitted.

That said, I think your problem is setting the RING count to 0. I should probably disallow this value as it makes no sense given the caller-id doesn't show up until after the 1st ring (usually).

Edit: Setting RING count to 0 or 1 is equivalent. The phone is actually not picked up until we have a caller-id record. If you are not getting the caller-id the phone will keep ringing. (More log data will help here.)

**Comment by thess** [comment]
At the risk of being redundant...https://github.com/thess/callattendant/issues/6#issuecomment-1881840818

> "Permit next call" feature is just what it was intended for. Implementing a different mechanism with a specified timeout (or not) makes sense in the context you describe. You can always blacklist the number of the caller who can follow directions even if their intent is not what you think.

Your use-case example of a timeout on the Permit Next Call feature is not necessarily the right action. I think you are proposing a different feature by which you want to wait for a **specific** caller who is expected to call back before some time-out for **that** caller only. The current screening implementation will "whilelist" a caller who presses '0' so their callback will work. If you didn't want to talk to them, you will need to remove them from the Whitelist and optionally, add to Blacklist. What doesn't work with your proposal is as follows:
* Caller A presses # (starts callback timer) and hangs up.
* Caller B calls and rings thru and is not screened and clears permit next call flag.
* Caller A calls and if not busy, gets screened again.

**Comment by whitedavidp** [comment]
You are certainly correct that the use case you mention can happen. I guess I have to wonder about how frequently it might happen during a relatively short interval?

I do think your idea to whitelist for a specific interval would be far better than my hair-brained, one-off. It would be more specific and not interfere with the regular permit next call mechanism.

And with the current Permit Next Call, is there a way to turn that off? For example, say I am doing a 2-step verification and I press the button on the web UI to do this. But the call does not come within a reasonable period. That might happen, for example, if the bank has the wrong number on file. So how does a normal user dis-engage it? The semaphore can be manually deleted, of course, and that should work. But if that isn't done, how does a user cancel the permit next request?

Thanks for all of your time and patience with this. I will look for the screener_greeting.wav to make it to github.

**Comment by thess** [comment]
When "Permit Next Call" is enabled, the button in the UI changes to "Permit Next Call (ON)", press button to turn it off -- It's a toggle.

**Comment by whitedavidp** [comment]
Sorry to have pulled things off-topic. My original concern was with delays in the modem actually getting off-hook when a call is being screened.

Right now, I am just trying to get the whitelist on 0 to work. I have gotten and configured the new screened_greeting.wav file. Thanks. And I have configured:

SCREENED_ACTIONS = ("answer", "greeting", "voice_mail")
SCREENED_GREETING_FILE = "screener_greeting.wav"
SCREENED_RINGS_BEFORE_ANSWER = 0

In this configuration, things basically "work" (or at least some of the time). What I don't understand is why the phone keeps ringing so many times in this configuration. I really want to hear at most 1 ring if I can. This seems work work when a call is blocked. I have both rings_before_answer set to 0. Based upon your above comment, I also tried changing the screened rings_before_answer to 1. But that did not improve things (nor would I expect it to do so based upon what little I understand at this point).

I have attached the console log of 2 calls coming in as well as the modem log file. I did have to add a line to call_handler in modem.py to get newlines in this output making it more readable:

if dev_mode:
      logfile.write(modem_data.encode())
      logfile.write(os.linesep.encode())
      logfile.flush()

Sometimes I get only 1 ring when screening, but most of the time 2 or even more.

Thanks

[callattendant.log](https://github.com/thess/callattendant/files/13968574/callattendant.log)
[modem.log](https://github.com/thess/callattendant/files/13968575/modem.log)

**Comment by whitedavidp** [comment]
Here is a console log showing what I mean. Please note the RINGs that follow the message "going off hook" on line 29 (there are 3). You will note that I have inserted sending the modem 'ATH1' before anything else in an effort to get the line off-hook more quickly. Sadly, that does not seem to make any difference. Also, I have added a core to the virtual box that is running Ubuntu/callattendant in hopes that it might help speed this up a bit. The virtual box certainly seems more snappy to me now. But it has not really done anything about this problem. I have also tried to change the nice-ness of the callattendant process without any change. Thanks!

Jan 17 14:52:18 callattendant[2699]: Waiting for next call...
Jan 17 14:52:32 callattendant[2699]: RING
Jan 17 14:52:33 callattendant[2699]: DATE=0117
Jan 17 14:52:33 callattendant[2699]: TIME=1452
Jan 17 14:52:33 callattendant[2699]: NMBR=xxxxxxxxxx
Jan 17 14:52:33 callattendant[2699]: NAME=WIRELESS CALLER
Jan 17 14:52:33 callattendant[2699]: > Queueing call xxxxxxxxxx for processing
Jan 17 14:52:33 callattendant[2699]: Adding to caller queue:
Jan 17 14:52:33 callattendant[2699]: {'DATE': '0117',
Jan 17 14:52:33 callattendant[2699]:  'NAME': 'WIRELESS CALLER',
Jan 17 14:52:33 callattendant[2699]:  'NMBR': 'xxxxxxxxxx',
Jan 17 14:52:33 callattendant[2699]:  'TIME': '1452'}
Jan 17 14:52:33 callattendant[2699]: Incoming call from xxxxxxxxxx
Jan 17 14:52:33 callattendant[2699]: > Checking next-call-permitted
Jan 17 14:52:33 callattendant[2699]: > Checking whitelist(s)
Jan 17 14:52:33 callattendant[2699]: >> Checking permitted patterns...
Jan 17 14:52:33 callattendant[2699]: > Checking blacklist(s)
Jan 17 14:52:33 callattendant[2699]: >> Checking blocked patterns...
Jan 17 14:52:33 callattendant[2699]: Caller has been screened
Jan 17 14:52:33 callattendant[2699]: > New call log entry #78
Jan 17 14:52:33 callattendant[2699]: ['WIRELESS CALLER',
Jan 17 14:52:33 callattendant[2699]:  'xxxxxxxxxx',
Jan 17 14:52:33 callattendant[2699]:  'Screened',
Jan 17 14:52:33 callattendant[2699]:  'Not found',
Jan 17 14:52:33 callattendant[2699]:  '17-Jan',
Jan 17 14:52:33 callattendant[2699]:  '02:52 PM',
Jan 17 14:52:33 callattendant[2699]:  '2024-01-17 14:52:33']
Jan 17 14:52:33 callattendant[2699]: --> xxxxxxxxxx Screened: Not found
Jan 17 14:52:33 callattendant[2699]: > Going off hook...
Jan 17 14:52:38 callattendant[2699]: RING
Jan 17 14:52:44 callattendant[2699]: RING
Jan 17 14:52:50 callattendant[2699]: RING
Jan 17 14:53:25 callattendant[2699]: >>> Lock acquired in pick-up()
Jan 17 14:53:25 callattendant[2699]: _send_and_read('ATH1','OK',5)
Jan 17 14:53:25 callattendant[2699]: '\r\n'
Jan 17 14:53:25 callattendant[2699]: 'OK\r\n'
Jan 17 14:53:25 callattendant[2699]: _send_and_read('AT+FCLASS=8','OK',5)
Jan 17 14:53:25 callattendant[2699]: '\r\n'
Jan 17 14:53:25 callattendant[2699]: 'OK\r\n'
Jan 17 14:53:25 callattendant[2699]: _send_and_read('AT+VSD=128,0','OK',5)
Jan 17 14:53:25 callattendant[2699]: '\r\n'
Jan 17 14:53:25 callattendant[2699]: 'OK\r\n'
Jan 17 14:53:25 callattendant[2699]: _send_and_read('AT+VLS=1','OK',5)
Jan 17 14:53:25 callattendant[2699]: '\r\n'
Jan 17 14:53:25 callattendant[2699]: 'OK\r\n'
Jan 17 14:53:25 callattendant[2699]: >> Playing greeting...
Jan 17 14:53:25 callattendant[2699]: > Playing /home/davidw/.callattendant/notifications/screener_greeting.wav...
Jan 17 14:53:25 callattendant[2699]: _send_and_read('AT+FCLASS=8','OK',5)
Jan 17 14:53:25 callattendant[2699]: '\r\n'
Jan 17 14:53:25 callattendant[2699]: 'OK\r\n'
Jan 17 14:53:25 callattendant[2699]: _send_and_read('AT+VSM=128,8000','OK',5)
Jan 17 14:53:25 callattendant[2699]: '\r\n'
Jan 17 14:53:25 callattendant[2699]: 'OK\r\n'
Jan 17 14:53:25 callattendant[2699]: _send_and_read('AT+VLS=1','OK',5)
Jan 17 14:53:25 callattendant[2699]: '\r\n'
Jan 17 14:53:25 callattendant[2699]: 'OK\r\n'
Jan 17 14:53:26 callattendant[2699]: _send_and_read('AT+VTX','CONNECT',5)
Jan 17 14:53:26 callattendant[2699]: '\r\n'
Jan 17 14:53:26 callattendant[2699]: 'CONNECT\r\n'
Jan 17 14:53:30 callattendant[2699]: [23B blob data]
Jan 17 14:53:30 callattendant[2699]: >> DCE Notification: <DLE>d
Jan 17 14:53:33 callattendant[2699]: [23B blob data]
Jan 17 14:53:33 callattendant[2699]: >> DCE Notification: <DLE>d
Jan 17 14:53:36 callattendant[2699]: [23B blob data]
Jan 17 14:53:36 callattendant[2699]: >> DCE Notification: <DLE>d
Jan 17 14:53:39 callattendant[2699]: [23B blob data]
Jan 17 14:53:39 callattendant[2699]: >> DCE Notification: <DLE>d
Jan 17 14:53:41 callattendant[2699]: [27B blob data]
Jan 17 14:53:42 callattendant[2699]: '\r\n'
Jan 17 14:53:42 callattendant[2699]: 'OK\r\n'
Jan 17 14:53:42 callattendant[2699]: >> Starting voice mail...
Jan 17 14:53:42 callattendant[2699]: MSG LED blinking
Jan 17 14:53:42 callattendant[2699]: > Waiting for key-press...
Jan 17 14:53:42 callattendant[2699]: _send_and_read('AT+FCLASS=8','OK',5)
Jan 17 14:53:42 callattendant[2699]: '\r\n'
Jan 17 14:53:42 callattendant[2699]: 'OK\r\n'
Jan 17 14:53:42 callattendant[2699]: _send_and_read('AT+VSD=128,100','OK',5)
Jan 17 14:53:42 callattendant[2699]: '\r\n'
Jan 17 14:53:42 callattendant[2699]: 'OK\r\n'
Jan 17 14:53:42 callattendant[2699]: _send_and_read('AT+VLS=1','OK',5)
Jan 17 14:53:42 callattendant[2699]: '\r\n'
Jan 17 14:53:42 callattendant[2699]: 'OK\r\n'
Jan 17 14:53:45 callattendant[2699]: [20B blob data]
Jan 17 14:53:49 callattendant[2699]: [21B blob data]
Jan 17 14:53:49 callattendant[2699]: [22B blob data]
Jan 17 14:53:49 callattendant[2699]: Exiting wait_for_keypress(8): Busy Tone... Aborting.
Jan 17 14:53:49 callattendant[2699]: Message Event triggered
Jan 17 14:53:49 callattendant[2699]: Resetting Message Indicator to show 0 unplayed messages
Jan 17 14:53:49 callattendant[2699]: MSG LED turned off
Jan 17 14:53:49 callattendant[2699]: MSG count: 0
Jan 17 14:53:49 callattendant[2699]: > Going on hook...
Jan 17 14:53:49 callattendant[2699]: _send_and_read('ATH0','OK',5)
Jan 17 14:53:50 callattendant[2699]: '\r\n'
Jan 17 14:53:50 callattendant[2699]: 'OK\r\n'
Jan 17 14:53:50 callattendant[2699]: >>> Lock released in hang-up()
Jan 17 14:53:50 callattendant[2699]: Waiting for next call...

and here is what is added to the modem.log:

RING
DATE=0117
TIME=1452
NMBR=xxxxxxxxxx
NAME=WIRELESS CALLER
RING
RING
RING

And BTW - even if I configure screening to be simply ("answer", ), I can see the same sort of behavior. Prior to moving to the new USR modem, I was using an ancient, external, caller-id capable modem and had commented out the more advanced modem stuff in the modem's pick_up() so that it just hangs up on everyone not permitted. With that modem and configuration, I never saw this behavior.

Thanks

**Comment by whitedavidp** [comment]
BTW - I decided to get around the virtualization of the serial port in the Ubuntu virtual box by registering the USR modem's USB device exclusively to the virtual box itself. With the virtual box running, the Windows host doesn't see it. In this way the modem shows up on /dev/ttyACM0 instead of on /dev/ttyS2. I had hoped that this might improve things. But so far, I see no difference. Rats!

**Comment by thess** [comment]
The problem as I now see it lies in the time-stamps between "Going off hook" and "Lock acquired...". That tells me the thread is stalled waiting for the serial port lock that the modem data handling thread may be holding too long or not releasing it as it should. I see a couple of problems with the code as it is currently structured but don't have a fix in mind yet.

You mention you are running on a Windows platform in a VM -- whose VM are you using? MS, VirtualBox, other? I don't think that to be an issue but as you mention serial line and USB devices can sometimes be a problem.

**Comment by thess** [comment]
One other thing - most important. Please run the application as supplied with a clean install (custom config OK). Do not ask me to debug any problems which may have been introduced by your local modifications. If you must, then I insist you supply your version for my review and/or testing.

**Comment by whitedavidp** [comment]
> The problem as I now see it lies in the time-stamps between "Going off hook" and "Lock acquired...". That tells me the thread is stalled waiting for the serial port lock that the modem data handling thread may be holding too long or not releasing it as it should. I see a couple of problems with the code as it is currently structured but don't have a fix in mind yet.

Wow, I didn't even look at that. That IS a long time!

> You mention you are running on a Windows platform in a VM -- whose VM are you using? MS, VirtualBox, other? I don't think that to be an issue but as you mention serial line and USB devices can sometimes be a problem.

This is Oracle Virtualbox. I have really not had problems with it. The same Ubuntu install runs a Heyu (home automation) server and a few other things (ssh, etc) but all of these are low use, not big performance hogs. Initially, I was allocating only a single core to Ubuntu's virtualbox. That always seemed sufficient up until now. But I have since added another. However, doing so clearly has not solved (or even impacted) this problem.

Heyu uses an on-motherboard serial port on /dev/ttyS1. It is set up as a host port and while I can see it in the Windows host, it cannot be accessed while the virtualbox runs. That seemed to work well enough so that is how I set up the new USR modem initially on /dev/ttyS2. Since then, I have assigned the modem via its USB device to the virtualbox and now the Windows host cannot see it at all when the virtualbox runs. It is now found at /dev/ttyACM0. Sadly, this clearly has not solved (or even impacted) this problem.

Thanks

**Comment by whitedavidp** [comment]
> One other thing - most important. Please run the application as supplied with a clean install (custom config OK). Do not ask me to debug any problems which may have been introduced by your local modifications. If you must, then I insist you supply your version for my review and/or testing.

Of course! This is a stock install (not yet updated to the most recently added stuff - except for the new .wav file for screening). It is true that I have tried to make point changes in an effort to see if I could get off-hook more rapidly. But when they have not helped, I reverted them. I also made (and have kept so far) the single modification to modem.py so that the modem's location has been hard-coded. Other than that, this is stock.

Truth is, I am not at all sure how to best update my current, "working" install to use the files you updated here. One thing I see is that requirements.txt has been fairly dramatically changed over the one I have from an install less than a week old. This implies some serious changes to the supporting stuff and I am not sure how to get that stuff updated either.

Thanks for all of your time and help on this. I feel so close - but still so far :-)

**Comment by whitedavidp** [comment]
Despite over 30 years as a software engineer, I know almost nothing about Python. But I have seen many references online to the performance of Python threads vs processes and the problems called by the GIL. In general, it appears (from what I have read) that CPU bound threads can be more of a problem than IO bound threads. At a first guess, I would suspect that what I am seeing is caused by IO but I am really too ignorant to know here.

**Comment by thess** [comment]
I don't think the problem here is the GIL. There is nothing compute bound in the app. I do suspect the modem serial input thread may not yield to the pending event delivery in the main call handling thread. I have always suspected the modem input implementation using `serial.readline()`, as originally designed, was going to need attention. I dislike long blocking timeouts in a loop as a programming paradigm. A quick test to see if indeed this is the case try the attached patch to `modem.py`. This is not the solution I am actually considering, but it will tell me if I am on the right track.

[test_thread_yield.patch.txt](https://github.com/thess/callattendant/files/13979712/test_thread_yield.patch.txt)

Also - You shouldn't need to set `ENV = "development"` except for your own curiosity. I don't need `modem.log`

WRT your install/update quandary, I will post some help here in a bit.
BTW - I'll see your 30 and raise you another 30 (almost).

**Comment by whitedavidp** [comment]
Thanks Ted! I was not sure if the patch was from the newest (which I don't yet have) or the previous version (which I do still have). So I hand edited the 2 changes into place.

I have to say that the difference is immediately startling! I have had 3 screened calls in a row all get handled on the 1st ring! Wow, that is impressive! And what an improvement!

BTW - I stopped counting years as an engineer when I retired from paying work. And I don't really want to think about all of those non-paying years :-) I obviously still hack-around with this stuff.

I look forward to being able to install/update without perturbing what I have too much.

Finally, for some reason unknown, leaving the USR modem registered to the virtualbox as a USB device creating /dev/ttyACM0 has stopped working and I get console errors that the port cannot be opened. It MAY be that this occurs only after a stop/start cycle once the virtualbox has been started (ie, after virtualbox startup it MAY be fine). For right now, I have gone back to using the Windows host COM port on /dev/ttyS2. And my results described above are from that configuration.

Best!

Thanks so much for all of your time and effort in this.

**Comment by whitedavidp** [comment]
Regards the USR modem when connected as a USB device in the virtualbox. I can confirm that the pattern of failure is:

(a) Start the virtualbox and it is found.
(b) Stop callattendant (I have it running as a service so to terminate it I use stop-callattendant from the desktop icon).
(c) Start callattendant (using start-callattendant from the desktop icon)

And no matter how many times I do (b), (c) the modem is no longer opened/used. I wonder if there is something that fails to fully close the port during shutdown?

Again, this does not seem to happen when using the Windows host COM port on /dev/ttyS2.

**Comment by thess** [comment]
Sorry, I've never tried running inside a VM using a serial USB device -- perhaps sometime in the future. I'm assuming you do have the VBox Guest Additions installed in your VM.

Some installation hints:

You can always install the latest bits directly from github
> \# Using pip
$ source venv/bin/activate
$ pip3 install "callattendant@git+https://github.com/thess/callattendant"

Don't be afraid to delete the installed package from `venv/lib/python3.x/site-packages`
>$ rm -rf venv/lib/python3.x/site-packages/callattendant*

Python setup is pretty resilient and will (usually) restore anything missing. You can also remove a package using `pip uninstall <pkg-name>` with the venv activated if installed with `pip`.

Or... you can wait until I spin up a new "wheel".

**Comment by whitedavidp** [comment]
Thanks. I will start working up to doing an update. With all things "working" now, I gotta take a break for a while and catch up on some other chores.

I do think I have the guest additions installed in my Ubuntu box. To the best of my knowledge I am not using them that much. I tried to install them into the virtualbox I have running Android x86 but that just doesn't work.

I need to read up on these wheels. As I said, I know next to nothing about Python.

Best and thanks again! These results are great. And I have cobbled-together some stuff to display and control some aspects of callattendant from my Android phone's home screen. It is basic but it works well enough for now and adds some convenience.
